lib/sax.js

```diff
@@ -538,86 +538,228 @@ function _copy(source, target) {
+				return errorHandler.fatalError('No blanks between DOCTYPE and document name');
+			}
+
+			// Parse the DOCTYPE name
+			doctype.name = getMatch(grammar.Name);
+			if (!doctype.name) return errorHandler.fatalError('DOCTYPE without name');
+			skipBlanks();
+
+			// Check for External ID and SystemId
+			if (compare('PUBLIC')) {
+				skip(6);
+				if (skipBlanks() < 1) {
+					return errorHandler.fatalError('No blanks between PUBLIC and public identifier');
 				}
-				var lastMatch = matchs[len - 1];
-				domBuilder.startDTD(name, pubid, sysid);
-				domBuilder.endDTD();
+				doctype.publicId = getMatch(grammar.PubidLiteral);
+				if (!doctype.publicId) return errorHandler.fatalError('PUBLIC without public identifier');
+				if (skipBlanks() < 1) {
+					return errorHandler.fatalError('No blanks between public identifier and SystemId');
+				}
+				doctype.systemId = getMatch(grammar.SystemLiteral);
+				if (!doctype.systemId) return errorHandler.fatalError('PUBLIC without SystemId');
+			} else if (compare('SYSTEM')) {
+				skip(6);
+				if (skipBlanks() < 1) {
+					return errorHandler.fatalError('No blanks between SYSTEM and SystemId');
+				}
+				doctype.systemId = getMatch(grammar.SystemLiteral);
+				if (!doctype.systemId) return errorHandler.fatalError('SYSTEM without SystemId');
+			}
+
+			// Parse internal subset
+			skipBlanks();
 
-				return lastMatch.index + lastMatch[0].length;
+			if (char() === '[') {
+				skip(1);
+				var intSubsetStart = index;
+				skipBlanks();
+				while (index < source.length) {
+					var current = null;
+					// Only in external subset
+					// if (char() === '<' && char(1) === '!' && char(2) === '[') {
+					// 	parseConditionalSections();
+					// } else
+					if (char() === '<' && char(1) === '!') {
+						switch (char(2)) {
+							case 'E':
+								if (char(3) === 'L') {
+									current = getMatch(grammar.elementdecl);
+								} else if (char(3) === 'N') {
+									current = getMatch(grammar.EntityDecl);
+								}
+								break;
+							case 'A':
+								current = getMatch(grammar.AttlistDecl);
+								break;
+							case 'N':
+								current = getMatch(grammar.NotationDecl);
+								break;
+							case '-':
+								current = getMatch(grammar.Comment);
+								break;
+						}
+					} else if (char() === '<' && char(1) === '?') {
+						current = parsePI();
+					} else if (char() === '%') {
+						current = getMatch(grammar.PEReference);
+					} else {
+						return errorHandler.fatalError('Error detected in Markup declaration');
+					}
+					if (!current) {
+						return errorHandler.fatalError('Error in internal subset at position ' + index);
+					}
+					skipBlanks();
+					if (char() === ']') {
+						doctype.internalSubset = source.substring(intSubsetStart, index);
+						skip(1);
+						break;
+					}
+				}
+				if (doctype.internalSubset === null) {
+					return errorHandler.fatalError('Error in internal subset: no closing bracket found');
+				}
```

Comment:
I wonder why this is not happening before setting the value some lines earlier. There is only that one case in which it could be null, since it is initialized with undefined, right?

Reply:
Yes it should check if it is undefined not null, I think in the previous `parseDoctypeDecl` function I initialized it to null instead of undefined.